File: 043/multiply-strings.js
/**
 * @param {string} num1
 * @param {string} num2
 * @return {string}
 */
function multiply(num1, num2) {
    if (num1 == null || num2 == null) {
        return "";
    }
    let paper = [];
    let len = num1.length + num2.length;
    let _num1 = num1.split("");
    let _num2 = num2.split("");

    for (let i = 0; i < _num1.length; i++) {
        for (let j = 0; j < _num2.length; j++) {
            paper[len - (i + j + 2)] += (_num1[i] - '0') * (_num2[j] - '0');
        }
    }

    for (let i = 0; i < len - 1; i++) {
        paper[i + 1] += (paper[i] - paper[i] % 10) / 10;
        paper[i] %= 10;
    }

    let s = "";
    for (let i = len - 1; i > 0; i--) {
        if ("" == s && paper[i] == 0) {
            continue;
        }
        s += paper[i];
    }
    s += paper[0];

    return s;
}

File: 043/MultiplyStrings.scala
object MultiplyStrings {
  def multiply(num1: String, num2: String): String = {
    if (num1 == "0" || num2 == "0") {
      return "0"
    }

    var temp = Array.fill[Int](num1.length + num2.length)(0)
    for (i <- num1.indices) {
      for (j <- num2.indices) {
//        temp(i+j+1) += (char2Character(num1(i)) - char2Character('0')) * (char2Character(num2(j)) - char2Character('0'))
//        temp(i+j+1) += (num1(i).toInt - 49) * (num2(j).toInt - 49)
        temp(i+j+1) += num1(i).asDigit * num2(j).asDigit
      }
    }

    for (i <- temp.length-1 to 1 by -1) {
      temp(i-1) += temp(i) / 10
      temp(i) = temp(i) % 10
    }

    if (temp(0) == 0) {
      temp = temp.slice(1, temp.length)
    }

    return temp.mkString("")
  }

  def main(args: Array[String]): Unit = {
    val num1 = "123"
    val num2 = "456"
    println(multiply(num1, num2))
  }
}
